Code/Source/HumanWorker/NpcNavigatorComponent.cpp

```diff
@@ -131,17 +200,34 @@ namespace ROS2::HumanWorker
             bearingError = GetSignedAngleBetweenUnitVectors(RobotDirection, GoalDirection);
         }
 
-        const AZ::Vector3 GoalToRobot = RobotPosition - goal.m_position;
-        const float CrossTrackError = goal.m_direction.Cross(GoalToRobot).GetLength();
+        const AZ::Vector3 zeroZ = AZ::Vector3::CreateOne() - AZ::Vector3::CreateAxisZ(1);
 
-        return Speed{ .m_linear = maxSpeed.m_linear, .m_angular = bearingError * maxSpeed.m_angular - CrossTrackError * crossTrackFactor };
-    }
+        const AZ::Vector3 routeVector = (goal.m_position - startPosition) * zeroZ;
+        const AZ::Vector3 positionVector = (RobotPosition - startPosition) * zeroZ;
+        const AZ::Vector3 crossProduct = routeVector.Cross(positionVector);
+        const float crossTrackError = (crossProduct.GetLength() / routeVector.GetLength()) * (crossProduct.GetZ() > 0 ? 1 : -1);
 
-    WaypointConfiguration NpcNavigatorComponent::FetchWaypointConfiguration(AZ::EntityId waypointEntityId)
-    {
-        WaypointConfiguration waypointConfiguration;
-        WaypointRequestBus::EventResult(waypointConfiguration, waypointEntityId, &WaypointRequests::GetConfiguration);
-        return waypointConfiguration;
+        float targetAngular = bearingError * maxSpeed.m_angular - crossTrackError * crossTrackFactor;
+        float targetLinear = maxSpeed.m_linear;
+
+        // Check if the worker will walk in a circle around the goal
+        // If so increase the angular speed to avoid this
+        if (AZ::IsClose(targetLinear / (goal.m_position - RobotPosition).GetLength(), targetAngular, 0.1f))
```

Comment:
What if `goal.m_position == RobotPosition`?
What is the reasoning behind comparing linear and angular speeds? Aren't the two different?